arrow/parser.py

```diff
@@ -141,53 +144,66 @@ def parse_iso(self, datetime_string):
         ]
 
         if has_time:
-            # Z is ignored entirely because fromdatetime defaults to UTC in arrow.py
-            if datetime_string[-1] == "Z":
-                datetime_string = datetime_string[:-1]
 
             if has_space_divider:
                 date_string, time_string = datetime_string.split(" ", 1)
             else:
                 date_string, time_string = datetime_string.split("T", 1)
 
-            time_parts = re.split("[+-]", time_string, 1)
-            colon_count = time_parts[0].count(":")
+            time_parts = re.split(r"[\+\-Z]", time_string, 1, re.IGNORECASE)
+            # TODO: should we prevent mixing basic and extended formats? would need to ensure that dates, times, and timezones are in same format
+            time_colon_count = time_parts[0].count(":")
+
+            is_basic_time_format = time_colon_count == 0
+            tz_format = "Z"
 
-            is_basic_time_format = colon_count == 0
+            # use 'ZZ' token instead since tz offset is present in non-basic format
+            if len(time_parts) == 2 and ":" in time_parts[1]:
+                tz_format = "ZZ"
 
-            has_tz = len(time_parts) > 1
-            has_hours = len(time_parts[0]) == 2
-            has_minutes = colon_count == 1 or len(time_parts[0]) == 4
-            has_seconds = colon_count == 2 or len(time_parts[0]) == 6
-            has_subseconds = re.search("[.,]", time_parts[0])
+            time_components = self._TIME_RE.match(time_parts[0])
+
+            if time_components is None:
+                raise ParserError(
+                    "Invalid time component provided. Please specify a format or provide a valid time component in the basic or extended ISO 8601 time format."
+                )
+
+            hours, minutes, seconds, subseconds_sep, subseconds = (
+                time_components.groups()
+            )
+
+            has_tz = len(time_parts) == 2
+            has_hours = hours is not None
+            has_minutes = minutes is not None
+            has_seconds = seconds is not None
+            has_subseconds = subseconds is not None
+
+            time_sep = "" if is_basic_time_format else ":"
 
             if has_subseconds:
-                time_string = "HH:mm:ss{}S".format(has_subseconds.group())
+                time_string = "HH{time_sep}mm{time_sep}ss{subseconds_sep}S".format(
+                    time_sep=time_sep, subseconds_sep=subseconds_sep
+                )
             elif has_seconds:
-                time_string = "HH:mm:ss"
+                time_string = "HH{time_sep}mm{time_sep}ss".format(time_sep=time_sep)
             elif has_minutes:
-                time_string = "HH:mm"
+                time_string = "HH{time_sep}mm".format(time_sep=time_sep)
             elif has_hours:
```

Comment:
L191-194 is where coverage is failing.

Reply:
L194 and L191->L194 are not reached, which means that L194 and the branch from L191->L194 are never triggered. Thinking about it, I have no idea how to even get it to L194. There was never an `else` condition before our new `0.15.0` changes, but we added it in. I don't think it can possibly be reached since all formatting issues should be caught by the `_TIME_RE` match above it (notice that I use the same error message since they technically are the same error of not matching time strings).

Reply:
I changed the `elif has_hours` to an `else` case and removed the `ParserError`, similar to how it is in the [current `master` code base](https://github.com/crsmithdev/arrow/blob/master/arrow/parser.py#L108).